File: mode_7_demo/Mode7Demo.java
import javax.swing.*;
import java.awt.*;
import java.awt.event.*;
import java.awt.image.BufferedImage;
import java.awt.image.DataBufferInt;
import java.io.File;
import java.io.IOException;
import javax.imageio.ImageIO;

public class Mode7Demo extends JPanel implements ActionListener, KeyListener {
    private static final int WIDTH = 1280;
    private static final int HEIGHT = 720;
    private static final int HORIZON = HEIGHT / 3;
    private static final int TEX_SIZE = 1024;
    private static final int FP_SHIFT = 6;
    private static final int FP_ONE = 1 << FP_SHIFT; // 256
    private static final int FP_HALF = FP_ONE >> 1; // 128

    private final BufferedImage screenImage;
    private int[] screenPixels;
    private int[] texturePixels;
    private int[] texturePixels2;
    private int[] skyPixels;  // renamed for clarity (was sky)

    // Camera/Player state (fixed point where appropriate)
    private int camX = 50 << FP_SHIFT; // 50.0
    private int camY = 50 << FP_SHIFT; // 50.0
    private int camZ = 10 << FP_SHIFT; // 8.0   ← typo in comment fixed but value unchanged
    private int angle = 0; // in "fixed point radians" — we'll use lookup or approx
    private int fov = 320 << FP_SHIFT; // 320.0

    // Movement flags
    private boolean up = false, down = false, left = false, right = false;
    private boolean w = false, a = false, s = false, d = false;

    // Simple sin/cos table (0..255 representing 0..2π)
    private static final int[] SIN_TABLE = new int[256];
    private static final int[] COS_TABLE = new int[256];

    static {
        for (int i = 0; i < 256; i++) {
            double rad = (i / 256.0) * 2 * Math.PI;
            SIN_TABLE[i] = (int) Math.round(Math.sin(rad) * FP_ONE);
            COS_TABLE[i] = (int) Math.round(Math.cos(rad) * FP_ONE);
        }
    }

    public Mode7Demo() {
        JFrame frame = new JFrame("render");
        frame.setDefaultCloseOperation(JFrame.EXIT_ON_CLOSE);
        frame.setSize(WIDTH, HEIGHT);
        frame.setResizable(false);
        frame.add(this);
        frame.addKeyListener(this);
        frame.setVisible(true);

        screenImage = new BufferedImage(WIDTH, HEIGHT, BufferedImage.TYPE_INT_RGB);
        screenPixels = ((DataBufferInt) screenImage.getRaster().getDataBuffer()).getData();

        texturePixels = new int[TEX_SIZE * TEX_SIZE];
        texturePixels2 = new int[TEX_SIZE * TEX_SIZE];
        generateTexture();

        Timer timer = new Timer(16, this);
        timer.start();
    }

    private void generateTexture() {
        try {
            BufferedImage img = ImageIO.read(new File("map3.png"));
            texturePixels = convertTo256x256Array(img);
            BufferedImage img3 = ImageIO.read(new File("map2.png"));
            texturePixels2 = convertTo256x256Array(img3);

            BufferedImage img2 = ImageIO.read(new File("sky.jpg"));
            skyPixels = convertTo256x256Array(img2);   // now using same 1024×1024 resized format
        } catch (Exception e) {
            e.printStackTrace();
        }
    }

    public static int[] convertTo256x256Array(BufferedImage image) {
        BufferedImage resized = new BufferedImage(TEX_SIZE, TEX_SIZE, BufferedImage.TYPE_INT_ARGB);
        Graphics2D g = resized.createGraphics();
        g.drawImage(image, 0, 0, TEX_SIZE, TEX_SIZE, null);
        g.dispose();

        int[] pixels = new int[TEX_SIZE * TEX_SIZE];
        resized.getRGB(0, 0, TEX_SIZE, TEX_SIZE, pixels, 0, TEX_SIZE);
        return pixels;
    }

    @Override
    public void actionPerformed(ActionEvent e) {
        update();
        render();
        repaint();
    }

    private void update() {
        // Rotation (angle wraps 0..255)
        if (left) angle = (angle + 2) & 255;
        if (right) angle = (angle - 2) & 255;

        // Movement
        int speed = 2 << FP_SHIFT; // 2.0 in fixed point
        int dir = (angle + 64) & 255; // +90 degrees ≈ +64 in 256-step circle
        int dx = mulFP(COS_TABLE[dir], speed);
        int dy = mulFP(SIN_TABLE[dir], speed);

        if (up || w) {
            camX += dx;
            camY += dy;
        }
        if (down || s) {
            camX -= dx;
            camY -= dy;
        }
    }

    public static int get_uv(int screen_x, int screen_y, int player_x, int player_y, int player_z, int angle, int fov) {
        int sinA = SIN_TABLE[angle & 255];
        int cosA = COS_TABLE[angle & 255];
        int p = (screen_y - HORIZON) << FP_SHIFT; // distance from horizon
        int z = mulFPDiv(fov, player_z, p); // z = (fov * camZ) / p
        
        int offsetX = (screen_x - (WIDTH >> 1)) << FP_SHIFT; // signed
        int floorX = mulFPDiv(player_z, offsetX, fov);
        int floorY = player_z;
        // Rotate
        int worldX = mulFP(floorX, cosA) - mulFP(floorY, sinA);
        int worldY = mulFP(floorX, sinA) + mulFP(floorY, cosA);
        // Translate
        int texX = worldX + player_x;
        int texY = worldY + player_y;
        // Wrap texture coordinates (positive modulo)
        int u = (texX >> FP_SHIFT) & (TEX_SIZE - 1);
        int v = (texY >> FP_SHIFT) & (TEX_SIZE - 1);
        return (u<<16&0xFFFF0000)|(v);
    }

    private void render() {
        int sinA = SIN_TABLE[angle & 255];
        int cosA = COS_TABLE[angle & 255];

        // ────────────────────────────────────────────────
        // Sky rendering (simple cylindrical / rotated panorama)
        // ────────────────────────────────────────────────
        for (int y = 0; y <= HORIZON; y++) {
            // Simple panoramic sky: horizontal wrap + vertical stretch/compress
            int v = (y * TEX_SIZE / (HORIZON + 1)) & (TEX_SIZE - 1);

            for (int x = 0; x < WIDTH; x++) {
                // Rotate sky horizontally according to view angle
                int uOffset = 3*-(angle << 2) & (TEX_SIZE - 1);           // ×4 gives smoother turning
                int u = (x + uOffset) & (TEX_SIZE - 1);

                screenPixels[y * WIDTH + x] = skyPixels[v * TEX_SIZE + u];
            }
        }

        // ────────────────────────────────────────────────
        // Floor rendering (unchanged except variable rename)
        // ────────────────────────────────────────────────
        for (int y = HORIZON + 1; y < HEIGHT; y++) {
            int p = (y - HORIZON) << FP_SHIFT; // distance from horizon

            int z = mulFPDiv(fov, camZ, p); // z = (fov * camZ) / p

            for (int x = 0; x < WIDTH; x++) {
                int offsetX = (x - (WIDTH >> 1)) << FP_SHIFT; // signed

                int floorX = mulFPDiv(z, offsetX, fov);
                int floorY = z;

                // Rotate
                int worldX = mulFP(floorX, cosA) - mulFP(floorY, sinA);
                int worldY = mulFP(floorX, sinA) + mulFP(floorY, cosA);

                // Translate
                int texX = worldX + camX;
                int texY = worldY + camY;

                // Wrap texture coordinates (positive modulo)
                int u = (texX >> FP_SHIFT) & (TEX_SIZE - 1);
                int v = (texY >> FP_SHIFT) & (TEX_SIZE - 1);

				if( ((texturePixels[v * TEX_SIZE + u])>>24) !=0 )
                	screenPixels[y * WIDTH + x] = texturePixels[v * TEX_SIZE + u];
				else
                	screenPixels[y * WIDTH + x] = 0xDEADBEEF;
            }
        }
        /*for (int y = HORIZON + 1; y < HEIGHT; y++) {
            int p = (y - HORIZON) << FP_SHIFT; // distance from horizon

            int z = mulFPDiv(fov, camZ-(5<<FP_SHIFT), p); // z = (fov * camZ) / p

            for (int x = 0; x < WIDTH; x++) {
                int offsetX = (x - (WIDTH >> 1)) << FP_SHIFT; // signed

                int floorX = mulFPDiv(z, offsetX, fov);
                int floorY = z;

                // Rotate
                int worldX = mulFP(floorX, cosA) - mulFP(floorY, sinA);
                int worldY = mulFP(floorX, sinA) + mulFP(floorY, cosA);

                // Translate
                int texX = worldX + camX;
                int texY = worldY + camY;

                // Wrap texture coordinates (positive modulo)
                int u = (texX >> FP_SHIFT) & (TEX_SIZE - 1);
                int v = (texY >> FP_SHIFT) & (TEX_SIZE - 1);

				if( ((texturePixels2[v * TEX_SIZE + u])>>24) !=0 ){
					screenPixels[(y) * WIDTH + x] = texturePixels2[v * TEX_SIZE + u];
				}/*else{
					if(y<HEIGHT-1)
					screenPixels[(y) * WIDTH + x] = screenPixels[(y-1) * WIDTH + x];
				}* /
            }
            for (int x = 0; x < WIDTH; x++) {
            }
        }*/
    }

    // Fixed-point multiply (24.8 × 24.8 → 48.16 → take middle 32 bits → back to 24.8)
    private static int mulFP(int a, int b) {
        return (int) (((long) a * b) >> FP_SHIFT);
    }

    // Fixed-point multiply + divide: (a * b) / c
    private static int mulFPDiv(int a, int b, int c) {
        return (int) (((long) a * b) / c);
    }

    @Override
    protected void paintComponent(Graphics g) {
        super.paintComponent(g);
        g.drawImage(screenImage, 0, 0, null);
        g.setColor(Color.WHITE);
        g.drawString("WASD / Arrows to move & turn", 10, 20);
        g.drawString("X: " + (camX >> FP_SHIFT) + " Y: " + (camY >> FP_SHIFT), 10, 35);
    }

    @Override
    public void keyPressed(KeyEvent e) {
        switch (e.getKeyCode()) {
            case KeyEvent.VK_W: case KeyEvent.VK_UP:    up = w = true; break;
            case KeyEvent.VK_S: case KeyEvent.VK_DOWN:  down = s = true; break;
            case KeyEvent.VK_A: case KeyEvent.VK_LEFT:  left = a = true; break;
            case KeyEvent.VK_D: case KeyEvent.VK_RIGHT: right = d = true; break;
        }
    }

    @Override
    public void keyReleased(KeyEvent e) {
        switch (e.getKeyCode()) {
            case KeyEvent.VK_W: case KeyEvent.VK_UP:    up = w = false; break;
            case KeyEvent.VK_S: case KeyEvent.VK_DOWN:  down = s = false; break;
            case KeyEvent.VK_A: case KeyEvent.VK_LEFT:  left = a = false; break;
            case KeyEvent.VK_D: case KeyEvent.VK_RIGHT: right = d = false; break;
        }
    }

    @Override
    public void keyTyped(KeyEvent e) {}

    public static void main(String[] args) {
        new Mode7Demo();
    }
}

File: mode_7_demo/c/mode7_demo.c
#include "../space_invaders/hw_contract.h"
#include "../space_invaders/sw/shm_map.h"

#include <stdint.h>
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <math.h>

#define SHM_NAME "/pynq_fbmmio"

#define TEX_SIZE 1024
#define HORIZON (H / 3)
#define FP_SHIFT 6
#define FP_ONE (1 << FP_SHIFT)

#define RGBA_BYTES_PER_PIXEL 4
#define TEX_PIXELS (TEX_SIZE * TEX_SIZE)
#define TEX_BYTES (TEX_PIXELS * RGBA_BYTES_PER_PIXEL)

typedef struct {
    int cam_x;
    int cam_y;
    int cam_z;
    int angle;
    int fov;
} mode7_state_t;

static int32_t sin_table[256];
static int32_t cos_table[256];

static inline int mul_fp(int a, int b) {
    return (int)(((int64_t)a * (int64_t)b) >> FP_SHIFT);
}

static inline int mul_fp_div(int a, int b, int c) {
    if (c == 0) {
        return 0;
    }
    return (int)(((int64_t)a * (int64_t)b) / c);
}

static void init_trig_tables(void) {
    for (int i = 0; i < 256; i++) {
        double rad = ((double)i / 256.0) * 2.0 * M_PI;
        sin_table[i] = (int32_t)llround(sin(rad) * FP_ONE);
        cos_table[i] = (int32_t)llround(cos(rad) * FP_ONE);
    }
}

static int load_rgba_texture(const char *path, uint32_t *dst_argb) {
    FILE *f = fopen(path, "rb");
    if (!f) {
        return -1;
    }

    uint8_t *rgba = (uint8_t *)malloc(TEX_BYTES);
    if (!rgba) {
        fclose(f);
        return -1;
    }

    size_t got = fread(rgba, 1, TEX_BYTES, f);
    fclose(f);
    if (got != TEX_BYTES) {
        free(rgba);
        return -1;
    }

    for (int i = 0; i < TEX_PIXELS; i++) {
        uint8_t r = rgba[4 * i + 0];
        uint8_t g = rgba[4 * i + 1];
        uint8_t b = rgba[4 * i + 2];
        uint8_t a = rgba[4 * i + 3];
        dst_argb[i] = ((uint32_t)a << 24) | ((uint32_t)r << 16) | ((uint32_t)g << 8) | (uint32_t)b;
    }

    free(rgba);
    return 0;
}

static void generate_fallback_floor(uint32_t *tex) {
    for (int y = 0; y < TEX_SIZE; y++) {
        for (int x = 0; x < TEX_SIZE; x++) {
            int checker = ((x >> 5) ^ (y >> 5)) & 1;
            uint8_t c = checker ? 0xAA : 0x66;
            tex[y * TEX_SIZE + x] = 0xFF000000u | ((uint32_t)c << 16) | ((uint32_t)c << 8) | c;
        }
    }
}

static void generate_fallback_sky(uint32_t *tex) {
    for (int y = 0; y < TEX_SIZE; y++) {
        for (int x = 0; x < TEX_SIZE; x++) {
            int grad = 90 + (y * 120) / TEX_SIZE;
            int cloud = (((x * 31) ^ (y * 17) ^ (x * y)) & 0x3F);
            int blue = grad + (cloud >> 3);
            int green = (grad * 3) / 5 + (cloud >> 4);
            int red = green / 2;
            if (red > 255) red = 255;
            if (green > 255) green = 255;
            if (blue > 255) blue = 255;
            tex[y * TEX_SIZE + x] = 0xFF000000u | ((uint32_t)red << 16) | ((uint32_t)green << 8) | (uint32_t)blue;
        }
    }
}

static void update_state(mode7_state_t *s, uint32_t buttons) {
    if (buttons & BTN_LEFT) {
        s->angle = (s->angle + 2) & 255;
    }
    if (buttons & BTN_RIGHT) {
        s->angle = (s->angle - 2) & 255;
    }

    int speed = 2 << FP_SHIFT;
    int dir = (s->angle + 64) & 255;
    int dx = mul_fp(cos_table[dir], speed);
    int dy = mul_fp(sin_table[dir], speed);

    if (buttons & BTN_UP) {
        s->cam_x += dx;
        s->cam_y += dy;
    }
    if (buttons & BTN_DOWN) {
        s->cam_x -= dx;
        s->cam_y -= dy;
    }
}

static void render_frame(uint32_t *dst, const mode7_state_t *s, const uint32_t *floor_tex, const uint32_t *sky_tex) {
    for (int y = 0; y <= HORIZON; y++) {
        int v = (y * TEX_SIZE / (HORIZON + 1)) & (TEX_SIZE - 1);
        int u_offset = (-(s->angle << 2) * 3) & (TEX_SIZE - 1);

        int row = y * W;
        for (int x = 0; x < W; x++) {
            int u = (x + u_offset) & (TEX_SIZE - 1);
            dst[row + x] = sky_tex[v * TEX_SIZE + u];
        }
    }

    int sin_a = sin_table[s->angle & 255];
    int cos_a = cos_table[s->angle & 255];

    for (int y = HORIZON + 1; y < H; y++) {
        int p = (y - HORIZON) << FP_SHIFT;
        int z = mul_fp_div(s->fov, s->cam_z, p);

        int row = y * W;
        for (int x = 0; x < W; x++) {
            int offset_x = (x - (W >> 1)) << FP_SHIFT;

            int floor_x = mul_fp_div(z, offset_x, s->fov);
            int floor_y = z;

            int world_x = mul_fp(floor_x, cos_a) - mul_fp(floor_y, sin_a);
            int world_y = mul_fp(floor_x, sin_a) + mul_fp(floor_y, cos_a);

            int tex_x = world_x + s->cam_x;
            int tex_y = world_y + s->cam_y;

            int u = (tex_x >> FP_SHIFT) & (TEX_SIZE - 1);
            int v = (tex_y >> FP_SHIFT) & (TEX_SIZE - 1);

            uint32_t pcol = floor_tex[v * TEX_SIZE + u];
            if ((pcol >> 24) != 0) {
                dst[row + x] = pcol;
            } else {
                dst[row + x] = 0xDEADBEEFu;
            }
        }
    }
}

int main(void) {
    shm_ctx_t shm;
    if (shm_open_map(&shm, SHM_NAME) != 0) {
        return 1;
    }

    init_trig_tables();

    uint32_t *floor_tex = (uint32_t *)malloc(sizeof(uint32_t) * TEX_PIXELS);
    uint32_t *floor_tex_2 = (uint32_t *)malloc(sizeof(uint32_t) * TEX_PIXELS);
    uint32_t *sky_tex = (uint32_t *)malloc(sizeof(uint32_t) * TEX_PIXELS);
    if (!floor_tex || !floor_tex_2 || !sky_tex) {
        fprintf(stderr, "texture allocation failed\n");
        free(floor_tex);
        free(floor_tex_2);
        free(sky_tex);
        shm_close_unmap(&shm);
        return 1;
    }

    if (load_rgba_texture("map3.rgba", floor_tex) != 0) {
        generate_fallback_floor(floor_tex);
    }
    if (load_rgba_texture("map2.rgba", floor_tex_2) != 0) {
        generate_fallback_floor(floor_tex_2);
    }
    if (load_rgba_texture("sky.rgba", sky_tex) != 0) {
        generate_fallback_sky(sky_tex);
    }

    mode7_state_t state;
    state.cam_x = 50 << FP_SHIFT;
    state.cam_y = 50 << FP_SHIFT;
    state.cam_z = 10 << FP_SHIFT;
    state.angle = 0;
    state.fov = 320 << FP_SHIFT;

    uint32_t last_ack = shm.regs->swap_ack;

    while (1) {
        uint32_t buttons = shm.regs->buttons;
        if (buttons & BTN_QUIT) {
            break;
        }

        update_state(&state, buttons);

        uint32_t *back = shm_back_buffer_ptr(&shm);
        render_frame(back, &state, floor_tex, sky_tex);

        shm.regs->swap_request = 1;
        while (shm.regs->swap_ack == last_ack) {
        }
        last_ack = shm.regs->swap_ack;
    }

    shm.regs->prop_quit = 1;
    free(floor_tex);
    free(floor_tex_2);
    free(sky_tex);
    shm_close_unmap(&shm);
    return 0;
}
